vmsyscall/src/memory_map.rs

```diff
@@ -0,0 +1,317 @@
+                .start_frame_number
+                .cmp(&r.range.start_frame_number)
+            {
+                Ordering::Equal => {
+                    if region.range.end_frame_number < r.range.end_frame_number {
+                        // Case: (r = `r`, R = `region`)
+                        // ----rrrrrrrrrrr----
+                        // ----RRRR-----------
+                        r.range.start_frame_number = region.range.end_frame_number;
+                        self.add_region(region);
+                    } else {
+                        // Case: (r = `r`, R = `region`)
+                        // ----rrrrrrrrrrr----
+                        // ----RRRRRRRRRRRRRR-
+                        *r = region;
+                    }
+                }
+                Ordering::Greater => {
+                    if region.range.end_frame_number < r.range.end_frame_number {
+                        // Case: (r = `r`, R = `region`)
+                        // ----rrrrrrrrrrr----
+                        // ------RRRR---------
+                        let mut behind_r = *r;
+                        behind_r.range.start_frame_number = region.range.end_frame_number;
+                        r.range.end_frame_number = region.range.start_frame_number;
+                        self.add_region(behind_r);
+                        self.add_region(region);
+                    } else {
+                        // Case: (r = `r`, R = `region`)
+                        // ----rrrrrrrrrrr----
+                        // -----------RRRR---- or
+                        // -------------RRRR--
+                        r.range.end_frame_number = region.range.start_frame_number;
+                        self.add_region(region);
+                    }
+                }
+                _ => {
+                    // Case: (r = `r`, R = `region`)
+                    // ----rrrrrrrrrrr----
+                    // --RRRR-------------
+                    r.range.start_frame_number = region.range.end_frame_number;
+                    self.add_region(region);
+                }
+            }
+
+            return;
+        }
+        panic!(
+            "region {:x?} is not a usable memory region\n{:#?}",
+            region, self
+        );
+    }
+
+    /// Sort the `MemoryMap` by region index.
+    pub fn sort(&mut self) {
+        self.entries.sort_unstable_by(|r1, r2| {
+            if r1.range.is_empty() {
+                Ordering::Greater
+            } else if r2.range.is_empty() {
+                Ordering::Less
+            } else {
+                let ordering = r1
+                    .range
+                    .start_frame_number
+                    .cmp(&r2.range.start_frame_number);
+
+                if ordering == Ordering::Equal {
+                    r1.range.end_frame_number.cmp(&r2.range.end_frame_number)
+                } else {
+                    ordering
+                }
+            }
+        });
```

Comment:
This would be far more readable if you had a `Frame` newtype that implemented `PartialOrd` and then implemented `PartialOrd` on `Region`.